src/picongpu/include/plugins/kernel/CopySpecies.kernel

```diff
@@ -33,8 +33,104 @@ namespace picongpu
 
 using namespace PMacc;
 
+/** copy particle to single frame
+ *
+ * copy particle data which are stored in linked list and superCell hierarchy
+ * to a single frame.
+ */
+struct CopySpeciesToGlobal
+{
+    DataSpace<simDim> gridDim;
+    int threads;
 
+    CopySpeciesToGlobal(const DataSpace<simDim>& gridDim, const int threads) :
+    gridDim(gridDim), threads(threads)
+    {
+
+    }
+
+    /** copy particles
+     *
+     * @param counter scalar offset in `destFrame`
+     * @param destFrame single frame were all particles are copied in
+     * @param srcBox particle box were particles are read from
+     * @param filter filter to select particles
+     * @param gpuOffset gpu localOffset or to sliding window begin
+     * @param mapper mapper which describe the area were particles are copied from
+     */
+    template<class T_DestFrame, class T_SrcBox, class T_Filter, class T_Space, class T_Mapping>
+    void operator()(int& counter, T_DestFrame destFrame, T_SrcBox srcBox, T_Filter filter, T_Space gpuOffset, T_Mapping mapper)
+    {
+        #pragma omp parallel for
+        for (int linearBlockIdx = 0;
+             linearBlockIdx < gridDim.productOfComponents();
+             ++linearBlockIdx
+             )
+        {
+            DataSpace<simDim> blockIdx(DataSpaceOperations<simDim>::map(gridDim, linearBlockIdx));
+
+            using namespace PMacc::particles::operations;
+
+            typedef T_DestFrame DestFrameType;
+            typedef typename T_SrcBox::FrameType SrcFrameType;
+            typedef T_Mapping Mapping;
+            typedef typename Mapping::SuperCellSize Block;
+
+            SrcFrameType *srcFramePtr;
+            int localCounter;
+            int globalOffset;
+            const int particlePerFrame = PMacc::math::CT::volume<SuperCellSize>::type::value;
+            int storageOffset[particlePerFrame];
+
+            bool isValid;
+
+            const DataSpace<Mapping::Dim> block = mapper.getSuperCellIndex(blockIdx);
+            const DataSpace<Mapping::Dim> superCellPosition((block - mapper.getGuardingSuperCells()) * mapper.getSuperCellSize());
+            filter.setSuperCellPosition(superCellPosition);
+
+            srcFramePtr = &(srcBox.getFirstFrame(block, isValid));
+
+            while (isValid) //move over all Frames
+            {
+                localCounter = 0;
+                /* clear storageOffset array*/
+                for (int threadIdx = 0; threadIdx < particlePerFrame; ++threadIdx)
+                {
+                    storageOffset[threadIdx] = -1;
+                    PMACC_AUTO(parSrc, ((*srcFramePtr)[threadIdx]));
+                    /*count particle in frame*/
+                    if (parSrc[multiMask_] == 1 && filter(*srcFramePtr, threadIdx))
+                        storageOffset[threadIdx] = localCounter++;
+                }
+
+                /* atomic update with openmp*/
+                #pragma omp critical
```

Comment:
I skip this optimization for now because the full "kernel" needs `170ms` for `26*10^6` particles (https://github.com/ComputationalRadiationPhysics/picongpu/pull/908)